Login page

foopang opened the pull request:
This PR implemented the basic login functionality. Once login succeeded, JWT token will be stored in local storage and user will be redirected to homepage.

Please note the following changes:

- Updated `conduit.effects.http` method so that it can handle `post` requests
- Updated `conduit.api.fetch` method to accept `headers` parameter for later access to user authenticated pages via stored JWT token

I designed the form components can be re-usable as much as possible. There is `UserFormBase` can be shared among login form, registration form and profile form. Hope this makes sense to you. 

For the form validation, it is working as expected now. However, there is a little glitch with the input fields being out of focus when you start typing the first letter, I'm still seeing why.

foopang committed:
Login page

roman01la commented:
@foopang this input fields glitch is caused by using `nil` as the initial value of a field, use empty string instead `""`

roman01la commented on src/conduit/api.cljs:
```diff
@@ -17,22 +18,45 @@
+       (->xhr xhr/get {:query-params params
+                       :headers headers}))))
+
+(defn post
```
I think it might be a good idea to generalize `fetch` function to cover both `get` and `post` requests by supplying `:method` option with one of the possible request methods.

Also `Content-Type` header could be set automatically by providing `:type :json` (this can be used later if we need something else rather than only JSON)

I'm not sure, but maybe `httpurr` library can do this for us.

foopang replied:
@roman01la I will create another PR for that.

roman01la commented on src/conduit/components/forms.cljs:
```diff
@@ -0,0 +1,97 @@
+(ns conduit.components.forms
+  (:require [rum.core :as rum]
+            [clojure.string :refer [join]]
```
It's always better to require ns using `:as` directive by giving it a name which reflects that namespace e.g. `clojure.string :as cstr`. It increases readability because you can see immediately where this function comes from.

roman01la commented on src/conduit/components/forms.cljs:
```diff
@@ -0,0 +1,97 @@
+
+(def form-state (atom initial-form-state))
+
+(def email-regex #"[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?")
```
Closure Library has a lot of useful utility functions. There's `goog.format.EmailAddress` class that does email validation: [link to docs](https://google.github.io/closure-library/api/goog.format.EmailAddress.html). Take a look at it, maybe it's good enough for our purpose.

roman01la commented on src/conduit/components/login.cljs:
```diff
@@ -0,0 +1,19 @@
+       [:p.text-xs-center
+        [:a {:href "#/register"} "Need an account?"]]
+       (LoginForm r route params)
+       ]]]]
```
Formatting.`]]]]` should be in the end of the previous line. Check [Clojure Style Guide](https://github.com/bbatsov/clojure-style-guide) if you are not sure about any formatting rules.

roman01la commented on src/conduit/components/forms.cljs:
```diff
@@ -0,0 +1,97 @@
+                           (.preventDefault e)
+                           (when (validate! form-data form-errors validators)
+                             (citrus/dispatch! r :user :login {:email (:email @form-data)
+                                                               :password (:password @form-data)}))))]
```
It's better to `deref` once and extract required values instead of multiple `@form-data`

roman01la commented on src/conduit/components/forms.cljs:
```diff
@@ -0,0 +1,97 @@
+  [r route params]
+  (let [fields [{:key :email :placeholder "Email"}
+                {:key :password :placeholder "Password" :type "password"}]
+        validators {:email [[not-empty? "Please enter email"]
```
This is optional. It might be a good idea to extract validation functionality into a reusable mixin.
So it could be used like this for example:

```clojure
(rum/defcs MyForm <
  (mixins/form
    {:fields fields
     :validators validators})
  [state]
  (let [{:keys [fields errors values on-submit on-change]} (:form state)]
    [:form {:on-submit on-submit}
      (for [{:keys [key type]} fields]
        [:input {:type type
                 :value (get values key)
                 :on-change #(on-change key (.. % -target -value))}])]))
```

I'm not sure, this is just an idea which may turn out to be a bad, overgeneralized, one :)

roman01la commented on src/conduit/components/forms.cljs:
```diff
@@ -0,0 +1,97 @@
+        validators {:email [[not-empty? "Please enter email"]
+                            [email? "Invalid Email"]]
+                    :password [[not-empty? "Please enter password"]]}
+        buttons '[:button.btn.btn-lg.btn-primary.pull-xs-right "Sign in"]
```
Quotation is not needed here. The vector is plain Hiccup, that's fine.

roman01la commented on src/conduit/components/forms.cljs:
```diff
@@ -0,0 +1,97 @@
+(rum/defc UserFormBase <
+  {:will-unmount
+   (fn [{[r] :rum/args :as state}]
+     (reset! form-state initial-form-state)
```
Shouldn't `form-state` be a local state of the component instead of a global one?

roman01la commented on src/conduit/components/forms.cljs:
```diff
@@ -0,0 +1,97 @@
+(defn email? [v]
+  (not (nil? (re-matches email-regex (str v)))))
+
+(defn validate! [form-data form-errors validators]
```
Validation function should be much simpler, there's no need for an internal state. Given a list of validators you can map over the list, execute each validator and return the result. Validation results should be set to a local state of the from component so it can display them.

roman01la commented on src/conduit/components/root.cljs:
```diff
@@ -2,7 +2,8 @@
             [citrus.core :as citrus]
             [conduit.components.home :as home]
-            [conduit.components.article :refer [Article]]))
+            [conduit.components.article :refer [Article]]
```
Use `:as` directive, similar to `home` require here

roman01la commented on src/conduit/controllers/user.cljs:
```diff
@@ -0,0 +1,35 @@
+  {:state initial-state})
+
+(defmethod control :login [_ [{:keys [email password]}] _]
+  {:state initial-state
```
I suppose this is used to reset the state?

foopang committed:
Fix login page

foopang commented:
@roman01la Thanks for your feedback! I've fixed most of the issues according to your comments. And will try to fix the rest later.

roman01la commented:
@foopang Overall the code looks good to me except form validation part. There shouldn't be any global or intermediate state. Great job though!

foopang committed:
Form state mixin

foopang commented:
@roman01la I did what you suggested to moved form state and etc into reusable mixin in [this commit](https://github.com/roman01la/cljs-rum-realworld-example-app/pull/8/commits/26283a1c977677d260678279536476c979d5d80f). It comes out that the code has become much concise and cleaner!

foopang committed:
Refactor form validation function

foopang commented:
https://github.com/roman01la/cljs-rum-realworld-example-app/pull/8/commits/e5ecc1868b8831a8bb3b9c1f60bdaf080da93edf I've refactored the form validation function a bit to avoid internal state, and moved it over to `conduit.helpers.form` namespace. Hope it's better now.

roman01la commented on src/conduit/components/forms.cljs:
```diff
@@ -0,0 +1,67 @@
+    [:fieldset.form-group
+     [:input.form-control.form-control-lg
+      {:placeholder placeholder
+       :on-change #(reset! form-data (assoc @form-data input-key (.. % -target -value)))
```
This can be done with `swap!`: `(swap! form-data assoc input-key (.. % -target -value))`.
Also the input shouldn't know about any state, it should be a generic component which calls passed in handlers. Move state altering logic into the parent component.

```clojure
(Input {:on-change #(swap! form-data assoc input-key %)})

;; inside of the Input
:on-change #(on-change (.. % -target -value))
```

roman01la commented on src/conduit/components/forms.cljs:
```diff
@@ -0,0 +1,67 @@
+       :on-change #(reset! form-data (assoc @form-data input-key (.. % -target -value)))
+       :on-blur validate-input!
+       :on-key-down validate-input!
+       :on-key-up validate-input!
```
do we really need to run validation check for all of those three events `on-blur`, `on-key-down` and `on-key-up`?

roman01la commented on src/conduit/components/forms.cljs:
```diff
@@ -0,0 +1,67 @@
+       :on-key-down validate-input!
+       :on-key-up validate-input!
+       :type (or input-type :text)
+       :value (get-in (rum/react form-data) [input-key])}]
```
Should be `get`. And again: `Input` shouldn't know about any state. This particular case makes `Input` coupled with form component.

roman01la commented on src/conduit/components/forms.cljs:
```diff
@@ -0,0 +1,67 @@
+            [conduit.helpers.form :as form-helper]))
+
+(rum/defc InputErrors [errors input-key]
+  (when-let [input-errors (get-in errors [input-key])]
```
I think this check should be outside of the component since it will return `nil` in case when there's no `input-errors`, so there's no point in rendering a noop component

roman01la commented on src/conduit/components/forms.cljs:
```diff
@@ -0,0 +1,67 @@
+     (map (fn [e] [:li e]) input-errors)]))
+
+(rum/defc ServerErrors < rum/reactive [r]
+  (when-let [server-errors (rum/react (citrus/subscription r [:user :errors]))]
```
the same here about top-level check

roman01la replied:
also having a subscription here to specific set of data doesn't allow to reuse the component, it should be somewhere in parent component which is specific to `:user :errors`, maybe `LoginForm`?

roman01la commented on src/conduit/components/forms.cljs:
```diff
@@ -0,0 +1,67 @@
+  {:key-fn (fn [{input-key :input-key}] input-key)}
+  [{:keys [placeholder input-key input-type validators form-data form-errors]}]
+  (let [input-validators (select-keys validators [input-key])
+        validate-input! #(form-helper/validate! form-data form-errors input-validators)]
```
Not sure about having validation function in the input here. What if we want an input without validation logic? Validation function could be passed into the component as a usual handler for certain events like `on-key-down`, etc.

```clojure
(Input {:on-key-down validate})
```

roman01la commented on src/conduit/components/forms.cljs:
```diff
@@ -0,0 +1,67 @@
+       :on-key-up validate-input!
+       :type (or input-type :text)
+       :value (get-in (rum/react form-data) [input-key])}]
+     (InputErrors (rum/react form-errors) input-key)]))
```
the same about state

roman01la commented on src/conduit/components/forms.cljs:
```diff
@@ -0,0 +1,67 @@
+                                         (.preventDefault e)
+                                         (when (-> (form-helper/validate! form-data form-errors form-validators)
+                                                   form-helper/valid?)
+                                           (let [{:keys [email password]} (deref form-data)]
```
use `@` instead of `deref`

roman01la commented on src/conduit/mixins.cljs:
```diff
@@ -17,3 +17,24 @@
+  {:will-mount
+   (fn [{[r _ params] :rum/args
+         :as state}]
+     (let [initial-form-state {:data (reduce #(assoc %1 %2 "") {} (map #(:key %) fields))
```
Use `:key` as a function instead of `#(:key %)`.
Also use `->>` macro for readability: `(->> fields (map :key) (reduce #(assoc %1 %2 "") {}))`

roman01la commented on src/conduit/mixins.cljs:
```diff
@@ -17,3 +17,24 @@
+   (fn [{[r _ params] :rum/args
+         :as state}]
+     (let [initial-form-state {:data (reduce #(assoc %1 %2 "") {} (map #(:key %) fields))
+                               :errors (reduce #(assoc %1 %2 nil) {} (map #(:key %) fields))}
```
the same as above

roman01la commented on src/conduit/mixins.cljs:
```diff
@@ -17,3 +17,24 @@
+                               :errors (reduce #(assoc %1 %2 nil) {} (map #(:key %) fields))}
+           form-state (atom initial-form-state)
+           form-data (rum/cursor-in form-state [:data])
+           form-errors (rum/cursor-in form-state [:errors])]
```
Since `form-state` is not used anywhere else and both cursors are assigned to component's state below here, it might be better to create a separate atoms for both `form-data` and `form-errors` instead

roman01la commented on src/conduit/mixins.cljs:
```diff
@@ -17,3 +17,24 @@
+              :form-data form-data
+              :form-errors form-errors
+              :initial-form-state initial-form-state)))
+   :will-unmount
```
Not sure if there's a need in resetting the state when component gets unmounted, since the state with the component itself will be lost there's no point in doing this, unless I'm missing something.

roman01la commented on src/conduit/helpers/form.cljs:
```diff
@@ -0,0 +1,25 @@
+(ns conduit.helpers.form
+  (:import goog.format.EmailAddress))
+
+(defn not-empty? [v]
```
looks like the entire function could be replaced with `empty?`

```clojure
(empty? nil) ;; true
(empty? "") ;; true
(empty? "  ") ;; false
```

roman01la commented on src/conduit/helpers/form.cljs:
```diff
@@ -0,0 +1,25 @@
+    :else true))
+
+(defn email? [v]
+  (.isValid (new EmailAddress v)))
```
it's common to use `(Constructor.)` syntax instead of `(new Constructor)`

```clojure
(EmailAddress. v)
```

roman01la commented on src/conduit/helpers/form.cljs:
```diff
@@ -0,0 +1,25 @@
+    @form-errors)
+
+(defn valid? [form-errors]
+  (reduce-kv #(if (not (empty? %3)) (reduced false) true) true form-errors))
```
use `if-not`

roman01la commented on src/conduit/helpers/form.cljs:
```diff
@@ -0,0 +1,25 @@
+(defn email? [v]
+  (.isValid (new EmailAddress v)))
+
+(defn validate! [form-data form-errors validators]
```
This function is still overly complex. It shouldn't know about any state (pass in data instead of atoms). Instead of doing side effects just transform and return data. There should not be any side effectful functions like `doseq`, `doall`, etc.

foopang committed:
Clean up the code and update form validation logic

foopang commented:
@roman01la Sorry, I have a little bit difficulty understanding all what you mean for the form validation part, but I tried hard refactoring it. I hope I get closer to what you want it to be like now in https://github.com/roman01la/cljs-rum-realworld-example-app/pull/8/commits/48504617aee7333fda8102938a6bc7b9ead986ed

roman01la commented:
@foopang LGTM, thanks! Going to merge this, maybe will do general cleanup afterwards.

roman01la commented:
@foopang I did refactor form validation logic, you can see it here 8d82e9c, just to get the idea of what I meant.

foopang commented:
OMG! Thanks! That's so awesome. Never thought it can be done like that.

roman01la commented:
@foopang It requires some knowledge of Rum's mixins, but now everything is happening inside of the mixin, which makes it easier to debug imho.

foopang commented:
@roman01la That's great! Just one some issue I've spotted, the submit form button seems no longer working even when it becomes clickable after all validation has passed.

roman01la commented:
@foopang fixed. thanks!
I suppose you are going to continue working on what's left in #6 ?

foopang commented:
Yes, will continue on it. Thanks!
